util/tracecpu/collector.go

```diff
@@ -0,0 +1,333 @@
+		}
+		entry.CPUTimeMsTotal += uint64(record.CPUTimeMs)
+	}
+
+	if len(ts.topSQLMap) <= ts.maxSQLNum {
+		return
+	}
+
+	// find the max CPUTimeMsTotal that should be evicted
+	digestCPUTimeList := make([]digestAndCPUTime, len(ts.topSQLMap))
+	{
+		i := 0
+		for key, value := range ts.topSQLMap {
+			data := digestAndCPUTime{
+				Key:            key,
+				CPUTimeMsTotal: value.CPUTimeMsTotal,
+			}
+			digestCPUTimeList[i] = data
+			i++
+		}
+	}
+	// QuickSelect will only return error when the second parameter is out of range
+	if err := quickselect.QuickSelect(digestAndCPUTimeSlice(digestCPUTimeList), ts.maxSQLNum); err != nil {
+		//	skip eviction
+		return
+	}
+	shouldEvictList := digestCPUTimeList[ts.maxSQLNum:]
+	for _, evict := range shouldEvictList {
+		delete(ts.topSQLMap, evict.Key)
+		sqlDigest, planDigest := decodeCacheKey(evict.Key)
+		ts.mu.Lock()
+		delete(ts.normalizedSQLMap, sqlDigest)
+		delete(ts.normalizedPlanMap, planDigest)
+		ts.mu.Unlock()
+	}
+}
+
+// RegisterNormalizedSQL registers a normalized sql string to a sql digest, while the former can be of >1M long.
+// The in-memory space for registered normalized sql are limited by TopSQL.normalizedSQLCapacity.
+//
+// This function should be thread-safe, which means parallelly calling it in several goroutines should be fine.
+// It should also return immediately, and do any CPU-intensive job asynchronously.
+// TODO: benchmark test concurrent performance
+func (ts *TopSQLCollector) RegisterNormalizedSQL(sqlDigest string, normalizedSQL string) {
+	ts.mu.RLock()
+	_, exist := ts.normalizedSQLMap[sqlDigest]
+	ts.mu.RUnlock()
+	if !exist {
+		ts.mu.Lock()
+		ts.normalizedSQLMap[sqlDigest] = normalizedSQL
+		ts.mu.Unlock()
+	}
+}
+
+// RegisterNormalizedPlan is like RegisterNormalizedSQL, but for normalized plan strings.
+// TODO: benchmark test concurrent performance
+func (ts *TopSQLCollector) RegisterNormalizedPlan(planDigest string, normalizedPlan string) {
+	ts.planRegisterChan <- &planRegisterJob{
+		planDigest:     planDigest,
+		normalizedPlan: normalizedPlan,
+	}
+}
+
+// this should be the only place where the normalizedPlanMap is set
+func (ts *TopSQLCollector) registerNormalizedPlanWorker(planDecoder planBinaryDecodeFunc) {
+	// NOTE: if use multiple worker goroutine, we should make sure that access to the digest map is thread-safe
+	for {
+		job := <-ts.planRegisterChan
+		ts.mu.RLock()
+		_, exist := ts.normalizedPlanMap[job.planDigest]
+		ts.mu.RUnlock()
+		if exist {
+			continue
+		}
+		planDecoded, err := planDecoder(job.normalizedPlan)
+		if err != nil {
+			fmt.Printf("decode plan failed: %v\n", err)
+			continue
+		}
+		ts.normalizedPlanMap[job.planDigest] = planDecoded
+	}
+}
+
+// snapshot will collect the current snapshot of data for transmission
+// This could run in parallel with `Collect()`, so we should guard it by a mutex.
+func (ts *TopSQLCollector) snapshot() []*tipb.CPUTimeRequestTiDB {
+	ts.mu.RLock()
```

Comment:
Is it possible that snapshot is implemented by swapping a atomic.Value? For example, we can save all current status in a `*Struct`. When doing the snapshot, we can first create a new `*Struct`, and then do an atomic swap with the current status, so that the snapshot is taken out atomically and we can process it later.